Use Velocity VarIntFrameDecoder

xism4 opened the pull request:
Velocity provides a pretty frame decoder that could be used without any problem in Waterfall (Also works on java 8)

After debating for some time, I want to continue improving Waterfall to the best of my knowledge. Since I work with it and many people are still using it. 
And if anyone wonders why if Velocity already has it? The same question should be asked with other patches that Waterfall uses from Velocity as well.

xism4 committed:
Use Velocity VarIntFrameDecoder

Kamillaova commented on BungeeCord-Patches/0066-Use-Velocity-VarintFrameDecoder.patch:
```diff
@@ -0,0 +1,213 @@
+         }
+ 
+-        in.markReaderIndex();
++        final VarintByteDecoder reader = new VarintByteDecoder();
```
allocating object on every decode, for what?

xism4 replied:
> 

https://github.com/PaperMC/Velocity/blob/dev/3.0.0/proxy/src/main/java/com/velocitypowered/proxy/protocol/netty/MinecraftVarintFrameDecoder.java#L41 (Velocity does same)

I have not wanted to modify the frame decoder, until there is a solid answer from the Waterfall developers, I should not change anything.

astei replied:
I specifically crafted `VarintByteDecoder` in such a fashion so that the HotSpot JVM would optimize away the object creation through inlining the code in `AbstractByteBuf.forEachByte()`, allowing escape analysis to do the rest of the work.

Appearances can be quite deceiving!

xism4 replied:
> Appearances can be quite deceiving!

I will left it like this then

Janmm14 replied:
> I specifically crafted `VarintByteDecoder` in such a fashion so that the HotSpot JVM would optimize away the object creation through inlining the code in `AbstractByteBuf.forEachByte()`, allowing escape analysis to do the rest of the work.
> 
> Appearances can be quite deceiving!

Inmy opinion its unneccessary fancyness youre doing there.

astei replied:
It's not fanciness, it was me observing the behavior of the JVM and guiding it in the best direction in optimizing the code.

Janmm14 replied:
So you say the overall shorter and in my opinion equally readable while loop is not getting to the same optimizations?

xism4 committed:
Replace GPL license with same as 0056 patch.

I saw that this patch (https://github.com/PaperMC/Waterfall/blob/master/BungeeCord-Patches/0056-Additional-DoS-mitigations.patch) mostly is catch-up exception system of Velocity for illegal connections, i try to replicate and fix license stuff.

xism4 committed:
Remove unused objects due trace-less catch

xism4 committed:
Re-implement TOO_BIG result

Kamillaova commented on BungeeCord-Patches/0066-Use-Velocity-VarintFrameDecoder.patch:
```diff
@@ -0,0 +1,188 @@
+ import java.util.List;
+ 
+-public class Varint21FrameDecoder extends ByteToMessageDecoder
+-{
```
https://github.com/PaperMC/Waterfall/blob/master/CONTRIBUTING.md

![image](https://user-images.githubusercontent.com/54859825/206477976-9cbffbd3-ed17-4d0f-bb40-06e9625767ca.png)

What is the reason for changing code style?

xism4 replied:
I followed imself Velocity's one, because it's simply easier to read and clean

Janmm14 replied:
This is an unneccessary change in a patch, the incentive is to keep patches simple so there is less merge conflict fixing needed.
This also means no import reordering or adding empty lines inbetween imports (ik its automatic by IDE usually).

xism4 replied:
> This is an unneccessary change in a patch, the incentive is to keep patches simple so there is less merge conflict fixing needed. This also means no import reordering or adding empty lines inbetween imports (ik its automatic by IDE usually).

I will follow that, thanks

Kamillaova commented on BungeeCord-Patches/0066-Use-Velocity-VarintFrameDecoder.patch:
```diff
@@ -0,0 +1,188 @@
+         }
+ 
+-        in.markReaderIndex();
++        final VarintByteDecoder reader = new VarintByteDecoder();
```
Again. I suggests replace it with something like

```java
int readVarint;
int bytesRead;
DecodeResult result = DecodeResult.TOO_SHORT;

var start = in.readerIndex;
var end = in.writerIndex() + in.getLength();

for (; start < end; ++start) {
  var byte = in.getByte(start); // yes, rangecheck ;<

  if (k == 0 && bytesRead == 0) {
    // tentatively say it's invalid, but there's a possibility of redemption
    result = DecodeResult.RUN_OF_ZEROES;
    continue;
  }
  if (result == DecodeResult.RUN_OF_ZEROES) {
    break;
  }
  readVarint |= (k & 0x7F) << bytesRead++ * 7;
  if (bytesRead > 3) {
    result = DecodeResult.TOO_BIG;
    break;
  }
  if ((k & 0x80) != 128) {
    result = DecodeResult.SUCCESS;
    break;
  }
  continue;
}
```

xism4 replied:
Sure, i can modify it.

Janmm14 replied:
Sounds kinda useless to use the DecodeResult at all. Instead of breaking and later switching over the enum, we should simply program the actions (exception throw) we'd take on each result immediately.

The if (run of zeroes) after the continue is also never being used.

And without you changing this, i woudln't mark this as resolved.

Janmm14 commented on BungeeCord-Patches/0066-Use-Velocity-VarintFrameDecoder.patch:
```diff
@@ -0,0 +1,188 @@
+-{
++public class Varint21FrameDecoder extends ByteToMessageDecoder {
+ 
+-    private static boolean DIRECT_WARNING;
```
The patch [0044-Don-t-use-a-bytebuf-for-packet-decoding.patch](https://github.com/PaperMC/Waterfall/blob/85c0a35f0b40a12ff0d7a35c6ea31287eecab707/BungeeCord-Patches/0044-Don-t-use-a-bytebuf-for-packet-decoding.patch) should be edited if you want to remove this line.

xism4 replied:
Got it

xism4 committed:
Merge branch 'PaperMC:master' into master

Janmm14 commented on BungeeCord-Patches/0066-Use-Velocity-VarintFrameDecoder.patch:
```diff
@@ -0,0 +1,188 @@
+--- /dev/null
++++ b/protocol/src/main/java/net/md_5/bungee/protocol/VarintByteDecoder.java
+@@ -0,0 +1,68 @@
++/*
++ * Courtesy of Velocity
++ * See https://github.com/VelocityPowered/Velocity/commit/5ceac16a821ea35572ff11412ace8929fd06e278
++ */
```
I see, this commit is before Velocity's implementation changed to GPL.

xism4 replied:
> I see, this commit is before Velocity's implementation changed to GPL.

Yes, i saw even on other patches of Waterfall same
